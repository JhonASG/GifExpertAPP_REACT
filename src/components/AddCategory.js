import React, {useState} from "react";
import PropsTypes from 'prop-types';

const AddCategory = ({setCategories}) =>{
    // Obtener el valor del input y voy a modificarle el mismo
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(cat => [inputValue, ...cat]);
            setInputValue('');
        }
    }

    return (
        /*En este caso no se necesita un fragment <></> porque con el <form></form> 
        agrupamos todos los elementos.*/
        <form onSubmit={handleSubmit}>
            <input type="text" value={inputValue} onChange={handleInputChange}/>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropsTypes.func.isRequired
}

export default AddCategory;