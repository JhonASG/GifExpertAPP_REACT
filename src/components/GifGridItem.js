import React from 'react'
import PropsTypes from "prop-types";

export const GifGridItem = ({title, url}) => {
	return (
		<>
			<div data-testid='container' className='card animate__animated animate__rotateInDownLeft'>
				<img src={url} alt={title} />
				<p>{title}</p>
			</div>
		</>
	)
}

GifGridItem.propTypes = {
	title: PropsTypes.string.isRequired,
	url: PropsTypes.string.isRequired
}