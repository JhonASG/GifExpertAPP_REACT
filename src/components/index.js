//Archivo de barril
export * from './AddCategory';
export * from './GifGridItem';
export * from './GifGrid';