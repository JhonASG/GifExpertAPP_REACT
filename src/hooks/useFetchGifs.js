/* todos los hooks por defecto empiezan con un use, esto es un estandar en react */
/* primer custom hook */
/* Me va a permitir saber cuando estoy cargando y cuando termino la cargada de la solicitud http */
import { useEffect, useState } from "react"
import { getGifs } from '../helpers/getGifs.js'

export const useFecthGifs = (category) =>{
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            });
        });
    }, [category]);

    return state; // regresa el objeto que tiene como parametros la data y el loading.
}