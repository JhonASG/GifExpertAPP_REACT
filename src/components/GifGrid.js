import React from 'react'
import useFetchGifs from '../hooks/useFetchGifs'
import {GifGridItem} from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
	/*const [images, setImages] = useState([]);

	// el segundo parametro indica las dependencias, solo se ejecuta el useEffect cuando surge la dependencia
	useEffect( () => {
		getGifs(category).then(imgs => setImages(imgs));
	}, [category]);*/

	// Custom Hook que al final es una funci贸n que return un estado
	const {data:images, loading} = useFetchGifs(category);

	return (
		<>
			<h3 className='animate__animated animate__bounce'>{ category }</h3>
			{loading && <p data-testid='carga' className='animate__animated animate__flash'>Loading ...</p>}
			<div className='card-grid'>
				{
					images.map(imgs => {
						// Si paso el parametro con el spread {... img} estoy enviando por separada cada una de las propiedades de la imagen (id, title, url)
						return <GifGridItem key={imgs.id} {... imgs} />
					})
				}
			</div>
			{/*
				Al presionar el boton lo que esta ocasionando es un cambio en el hook useState()
				por tal motivo al React detectar un cambio va a renderizar de nuevo la petici贸n
				http, lo que puede ocasionar un ciclo infinito que provoque un fallo en la app
				por tal motivo se utiliza useEffect

				useEffect -> condicionamos que un c贸digo se ejecuta siempre y cuando se cumpla una
				condici贸n dada.

				const {count, setCount} = useState(0);
				<button onClick={() => {setCount(count + 1)}}>Contador</button>
			*/}
		</>
	)
}

GifGrid.propTypes = {
	category: PropTypes.string.isRequired 
}