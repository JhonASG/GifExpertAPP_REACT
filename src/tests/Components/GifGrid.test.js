import React from "react";
import '@testing-library/jest-dom';
import { render, screen } from "@testing-library/react";
import { GifGrid } from "../../components/GifGrid";
import useFetchGifs from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe('Pruebas para el componente GifGrid', () => {
    const category = 'Goku';

    test('Debe renderizar correctamente el componente', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

		const view = render(<GifGrid category={category} />);

        const p = screen.queryByTestId('carga').isConnected;

        expect(view).toMatchSnapshot();
        expect(p).toBe(true);
    });

    // Las pruebas deben ser unitarias para evaluarlas por separado
    test('Debe mostrar items cuando se cargan imagenes del hook useFetchGifs', () => {
        const gifs = [{
            id: '123',
            title: "Goku",
            url: "https://Goku.jpg"
        },
        {
            id: '345',
            title: "Gohan",
            url: "https://Goku.jpg"
        }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

		const view = render(<GifGrid category={category} />);
        const p = screen.queryByTestId('carga');

        expect(view).toMatchSnapshot();
        expect(p).toBeNull();
        expect(screen.queryAllByTestId('container').length).toBe(gifs.length);
    });
});