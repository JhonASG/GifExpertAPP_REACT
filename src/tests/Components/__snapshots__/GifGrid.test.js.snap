// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para el componente GifGrid Debe mostrar items cuando se cargan imagenes del hook useFetchGifs 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="animate__animated animate__bounce"
      >
        Goku
      </h3>
      <div
        class="card-grid"
      >
        <div
          class="card animate__animated animate__rotateInDownLeft"
          data-testid="container"
        >
          <img
            alt="Goku"
            src="https://Goku.jpg"
          />
          <p>
            Goku
          </p>
        </div>
        <div
          class="card animate__animated animate__rotateInDownLeft"
          data-testid="container"
        >
          <img
            alt="Gohan"
            src="https://Goku.jpg"
          />
          <p>
            Gohan
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h3
      class="animate__animated animate__bounce"
    >
      Goku
    </h3>
    <div
      class="card-grid"
    >
      <div
        class="card animate__animated animate__rotateInDownLeft"
        data-testid="container"
      >
        <img
          alt="Goku"
          src="https://Goku.jpg"
        />
        <p>
          Goku
        </p>
      </div>
      <div
        class="card animate__animated animate__rotateInDownLeft"
        data-testid="container"
      >
        <img
          alt="Gohan"
          src="https://Goku.jpg"
        />
        <p>
          Gohan
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Pruebas para el componente GifGrid Debe renderizar correctamente el componente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="animate__animated animate__bounce"
      >
        Goku
      </h3>
      <p
        class="animate__animated animate__flash"
        data-testid="carga"
      >
        Loading ...
      </p>
      <div
        class="card-grid"
      />
    </div>
  </body>,
  "container": <div>
    <h3
      class="animate__animated animate__bounce"
    >
      Goku
    </h3>
    <p
      class="animate__animated animate__flash"
      data-testid="carga"
    >
      Loading ...
    </p>
    <div
      class="card-grid"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
